<a id='top'></a>

# CSCI3022 S21
# Homework 1: Measures of Centrality and Dispersion
***

**Name**: ________________________________________________________________________

***

This assignment is due on Canvas by **MIDNIGHT on Monday, February 1**. Your solutions to theoretical questions should be done in Markdown directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your classmates, but **you must write all code and solutions on your own**.

**NOTES**: 

- Any relevant data sets should be available on Canvas. To make life easier on the graders if they need to run your code, do not change the relative path names here. Instead, move the files around on your computer.
- If you're not familiar with typesetting math directly into Markdown then by all means, do your work on paper first and then typeset it later.  Here is a [reference guide](https://math.meta.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference) linked on Canvas on writing math in Markdown. **All** of your written commentary, justifications and mathematical work should be in Markdown.  I also recommend the [wikibook](https://en.wikibooks.org/wiki/LaTeX) for LaTex.
- Because you can technically evaluate notebook cells is a non-linear order, it's a good idea to do **Kernel $\rightarrow$ Restart & Run All** as a check before submitting your solutions.  That way if we need to run your code you will know that it will work as expected. 
- It is **bad form** to make your reader interpret numerical output from your code.  If a question asks you to compute some value from the data you should show your code output **AND** write a summary of the results in Markdown directly below your code. 
- 95 points of this assignment are in problems.  The remaining 5 are for neatness, style, and overall exposition of both code and text.
- This probably goes without saying, but... For any question that asks you to calculate something, you **must show all work and justify your answers to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit. 

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Problem 3](#p3) |
---

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
%matplotlib inline

[Back to top](#top)

<br>

<a/ id='p1'></a>

### (15 points) Problem 1: Theory and Computation (Means and Medians)
***

A method to investigate the sensitivity of the sample mean and sample median to extreme outliers and changes in the dataset is to replace one or more elements in a given dataset by a number $y$ and investigate the effect when $y$ changes. To illustrate this, consider the following dataset:
 
$$  12.2 \quad -5.0 \quad 1.0 \quad 3.8 \quad -4.1 \quad 5.9 \quad 1.9 \quad 9.0 \quad y$$




**Part A:** Compute the sample mean and sample median for $y=3.5$. Then compute both quantities again for $y=6$.

In [2]:
#Your Code here
def addNumbertoList(num):
    values = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9,0]
    newValues = values.copy()
    newValues.append(num)
    return newValues

print("The mean with y = 1.5 is: ", np.mean(addNumbertoList(1.5)))
print("The median with y = 1.5 is: ", np.median(addNumbertoList(1.5)))
print("The mean with y = 6 is: ", np.mean(addNumbertoList(6)))
print("The median with y = 6 is: ", np.median(addNumbertoList(6)))

The mean with y = 1.5 is:  2.62
The median with y = 1.5 is:  1.7
The mean with y = 6 is:  3.0700000000000003
The median with y = 6 is:  2.8499999999999996


**Part B:** Is there a value for $y$ that would make the mean of the data equal to 6? If so, calculate the value of $y$ that makes the mean equal to 6. If not, clearly explain why not.
    
Is there a value for $y$ that would make the median of the data equal to 6? If so, calculate the values of $y$ that makes the median equal to 6. If not, clearly explain why not.

In [3]:
#Your code for Part C, here

values = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9,0]
sumVals = np.sum(values)
N = len(values)+1

print("The number that will make the mean equal to 6 is: ", (6*N - (sumVals)))

print("The sorted data values excluding y are:",np.sort(values))

The number that will make the mean equal to 6 is:  35.3
The sorted data values excluding y are: [-5.  -4.1  0.   1.   1.9  3.8  5.9  9.  12.2]


There is **not** however a value for $y$ that will make the median equal to 6. Note the ordered list of data values excluding $y$ above.

The middle data value is 1.9. If we add a number $y$ such that $1 \leq y \leq 1.9, then the median will be an average of $y$ and 1.9, such that $1.45 \leq \tilde{x} \leq 1.9$. 

Likewise, if we add a number $y$ such that $1.9 \leq y \leq 3.8$, the median will be the average of $y$ and 1.9, this time ranging in values from $1.9 \leq \tilde{x} \leq 2.85 $.

Thus, the median, $\tilde{x}$ can never equal 6.


**Part C**: Compute the sample variance and the sample standard deviation for the original dataset given in part A, with $y=6$.

In [4]:
#Your code for Part C, here# Solution:

values_y6 = addNumbertoList(6)
print("The variance with y = 6: ", np.var(values_y6))
print("the std dev with y = 6: ", np.std(values_y6))

The variance with y = 6:  26.726099999999995
the std dev with y = 6:  5.169729199871111


**Part D**: Compute both the sample mean and median for the following cases: 
- $y=9$ 
- $y=50$ 
- $y=4.36$ 
- $y \to \infty$ 
- $y \to -\infty$

In [5]:
# SOLUTION:

yvals = [9,50,4.36,np.Infinity,-np.Infinity]
for y in yvals:
    x = np.array([y,4.2, 5.1, 5.0, 3.8, 4.1, 5.5, 1.9])
    print("When y is {}, the median is {}".format(y,np.median(x)))#Your code for Part D, here
    print("When y is {}, the mean is {}".format(y,np.mean(x)))#Your code for Part D, here

When y is 9, the median is 4.6
When y is 9, the mean is 4.824999999999999
When y is 50, the median is 4.6
When y is 50, the mean is 9.95
When y is 4.36, the median is 4.28
When y is 4.36, the mean is 4.245
When y is inf, the median is 4.6
When y is inf, the mean is inf
When y is -inf, the median is 4.15
When y is -inf, the mean is -inf


**Part E**: Think about the previous parts, above, and describe in words or mathematical notation the answers to the following two questions:

- By varying $y$, what is the set of all the possible values that the sample mean could take on?
- By varying $y$, what is the set of all the possible values that the sample median could take on? Specifically, for what sets of $y$ values does the median take on its different possible values?

***
**Solution:**

We can push the sample mean around a lot, since we have an equation above that tells us what to set $y$ to for any $\bar{x}$. So $\bar{x}$ can be anything. In math, we could write this as $(-\infty,\infty)$.

The median is different. When $y\geq 5.0$, the median will be $4.6$.  Similarly, whenever $y \leq 4.1$, the median will be $4.2$. If $4.1 < y < 5.0$, then the median will be the mean of 4.3 and $y$. Therefore, the median can be $\{4.28, 4.6,[4.1,5.0]\}$.
***

**Part F**: Describe in words or mathematical notation, what happens to the sample standard deviation when $y$ is varied in the following ways: 
 
- $y \to \infty$ 
- $y \to \bar{x}$

***
**Solution:**

As $y \to \infty$, the sample standard deviation $ \to \infty $. Since the sample standard deviation involves calculating the squared distance from each data value to the mean, by setting y equal to the mean, the standard deviation will decrease.
***

[Back to top](#top)
<a/ id='p2'></a>

## (15 pts) Problem 2: Theory and Computation (Scaling)
***
Consider the following 3 data sets:

`A=[0,1,2,3,4,5,6,7,8,9,10,11,12]`

`B=[0,0,0,12,7,18,47,25,0,13,0,35]`

`C` is the random data set generated by using `np.random.exponential(scale=43, size=1000)`

For each data set, perform the following computations in parts A, B, and C:

**Part A:** Compute and print the mean and standard deviation of the data set.

In [6]:
A = [0,1,2,3,4,5,6,7,8,9,10,11,12]
B = [0,0,0,12,7,18,47,25,0,13,0,35]
C = np.random.exponential(scale=43, size=1000)

print('The mean of A is: ', np.mean(A))
print('The std of A is: ', np.std(A,ddof=1))

print('The mean of B is: ', np.mean(B))
print('The std of B is: ', np.std(B,ddof=1))

print('The mean of C is: ', np.mean(C))
print('The std of C is: ', np.std(C,ddof=1))

The mean of A is:  6.0
The std of A is:  3.8944404818493075
The mean of B is:  13.083333333333334
The std of B is:  15.64061341120571
The mean of C is:  43.528127157640185
The std of C is:  43.1039795338583


**Part B:** Compute and print  the mean and standard deviation of the new data set formed by subtracting the original mean from each observation.

In [7]:
A_new = [xi-np.mean(A) for xi in A]
B_new = [xi-np.mean(B) for xi in B]
C_new = [xi-np.mean(C) for xi in C]

print('The mean of A_new is: ', np.mean(A_new))
print('The std of A_new is: ', np.std(A_new,ddof=1))

print('The mean of B_new is: ', np.mean(B_new))
print('The std of B_new is: ', np.std(B_new,ddof=1))

print('The mean of C_new is: ', np.mean(C_new))
print('The std of C_new is: ', np.std(C_new,ddof=1))

The mean of A_new is:  0.0
The std of A_new is:  3.8944404818493075
The mean of B_new is:  -1.1842378929335002e-15
The std of B_new is:  15.64061341120571
The mean of C_new is:  -3.865352482534945e-15
The std of C_new is:  43.1039795338583


**Part C:** Compute and print  the mean and standard deviation of the new data set formed by subtracting the original mean from each observation and then dividing by the original standard deviation.

In [8]:
#Solution:

A_newest = [xi/np.std(A, ddof=1) for xi in A_new]
B_newest = [xi/np.std(B, ddof=1) for xi in B_new]
C_newest = [xi/np.std(C, ddof=1) for xi in C_new]

print('The mean of A_newest is: ', np.mean(A_newest))
print('The std of A_newest is: ', np.std(A_newest,ddof=1))

print('The mean of B_newest is: ', np.mean(B_newest))
print('The std of B_newest is: ', np.std(B_newest,ddof=1))

print('The mean of C_newest is: ', np.mean(C_newest))
print('The std of C_newest is: ', np.std(C_newest,ddof=1))

The mean of A_newest is:  -5.124106267500723e-17
The std of A_newest is:  1.0
The mean of B_newest is:  -5.551115123125783e-17
The std of B_newest is:  0.9999999999999998
The mean of C_newest is:  -9.592326932761352e-17
The std of C_newest is:  1.0


**Part D:** Why might this result matter?

***
**Solution:**
Subtracting the mean and dividing by the standard deviation of the original data set creates a new data set with mean 0 and standard deviation equal to 1. This is how data is standardized and is important when comparing data sets that were recorded using different units. For example, perhaps we wanted to compare daily temperatures for the USA that were recorded in Fahrenheit with daily temperatures in Australia that were recorded in Celsius. Transforming each data set using the previously described method allows for better comparison between the data sets. 
***

**Part E:** Looking at each of the 3 data sets, come up with a real-world context where those spectrum of numbers might make for reasonable observations.

***
**Solution:**
Data in A are a list of small integers: this might be the kind of values we see if we're counting things (number of wins for a sports team, number of students getting specific grade values, etc.).

Data in B are mostly small integers, but have a lot of zeros: one common example of data with lots of zeros is precipitation data.  Another example might be feedings times of animals: 0's denote specimens that choose not to eat while the others have times recorded.

Data in C are a big collection of numbers from 0 to infinity, which high decimal precision.  We'll learn more applications for exactly those `exponential` points, but one that makes sense is recording times, or weights, or any other physical attribute where we tend to not observe zero or negative values.
***

**Part F:** Let's prove a generalization of the result you likely observed in part B.  What happens if you take data set and "recenter" it by adding or subtracting the same number $c$ to each and every observation?

In other words, show that the mean of the data set $Y$ given by $Y_i=X_i+c$ is $\bar{Y}=\bar{X}+c$.

(**Hint**: start with the definition of the mean: $$\bar{Y}=\frac{\sum_{i=1}^n Y_i}{n},$$ then work with the right hand side!)

$$\bar{Y}=\frac{\sum_{i=1}^n Y_i}{n},$$
$$=\frac{\sum_{i=1}^n X_i+c}{n}$$
$$=\frac{\sum_{i=1}^n X_i}{n}+\frac{\sum_{i=1}^n c}{n}$$
$$=\bar{X}+\frac{nc}{n}=\bar{X}+c$$


[Back to top](#top)
<a/ id='p3'></a>

## (15 pts) Problem 3: Computation (Streaming Means)
***

Data science is often divided into two categories: questions of *what* the best value might be to repreesnt a data problem, and questions of *how* to compute that data value.  Question 1 - and prior lectures - should tell you that computing the mean is valuable!  But *how* do we compute the mean?

Let $x_1, x_2, \ldots, x_n$ be $n$ observations of a variable of interest.  Recall that the sample mean $\bar{x}_n$ and sample variance $s^2_n$ are given by 
<a id='eq1'></a>
$$
\bar{x}_n = \frac{1}{n}\sum_{k=1}^n x_k \quad \textrm{and} \quad s^2_n = \frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x}_n\right)^2 \qquad \tag{Equation 1}
$$

**Part A**:

How many computations - floating point operations: addition, subtraction, multiplication, division each count as 1 operation - are required to compute the mean of the data set with $n$ observations?


**Solution:**

Computing a mean with $n$ observations takes $n-1$ additions followed by a division, for $n$ operations total.

**Part B**:

Now suppose our data is *streaming*- we slowly add observations one at a time, instead of seeing the entire data set at once.  We are still interested in the mean, so if we stream the data set `[4,6,0,10, ...]`, we first compute the mean of the the first data point `[4]`, then we recompute the mean of the first two points `[4,6]`, then we recompute the mean of three `[4,6,0]`, and so forth.

Suppose we recompute the mean from scratch after each and every one of our $n$ observations are one-by-one added to our data set.  How many floating point operations are spent computing (and re-computing) the mean of the data set?

**Solution:**

 This means the first mean takes 1, the second mean takes 2, the third mean takes 3, and so forth.  Once we're computed $n$ sequential means, the total amount of computations is:

$1+2+3+\dots + n= \sum_{i=1}^n i = \frac{n(n+1)}{2}$ computations.  This is *order* $n^2$.

We should be convinced that streaming a mean costs a lot more computer time than just computing once!

In this problem we explore a smarter method for such an _online_ computation of the mean.  

**Result**: The following relation holds between the mean of the first $n-1$ observations and the mean of all $n$ observations: 

$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$


A proof of this result is in the [Appendix](#Appendix) after this problem, and requires some careful manipulations of the sum $\bar{x}_n$.  Your task will be to computationally verify and utilize this result.

**Part C**: Write a function `my_sample_mean` that takes as its input a numpy array and returns the mean of that numpy array using the formulas from class ([Equation 1](#eq1)). Write another function `my_sample_var` that takes as its input a numpy array and returns the variance of that numpy array, again using the formulas from class ([Equation 1](#eq1)). You may **not** use any built-in sample mean or variance functions.

In [9]:
#Solution:

def my_sample_mean(xx):
    n = len(xx)
    total = np.sum(xx)
    return total/n

def my_sample_var(xx):
    mean = my_sample_mean(xx)
    centered = xx - mean
    squared = centered**2
    return np.sum(squared)/(len(xx)-1)

**Part D**: Use your functions from Part C to compute the sample mean and sample variance of the following array, which contains the minutes late that the BuffBus is running on Friday afternoon.

`bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]`

In [10]:
bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]
mean = my_sample_mean(bus)
var = my_sample_var(bus)
print("Sample mean is: ", mean,"; sample variance is: ", var)

Sample mean is:  56.416666666666664 ; sample variance is:  7274.628787878787


**Part E**: Implement a third function called `update_mean` that implements the formula discussed after part B. Note that this function will need to take as its input three things: $x_n$, $\bar{x}_{n-1}$ and $n$, and returns $\bar{x}_{n}$. A function header and return statement are provided for you. This function may be auto-graded, so please do not change the given header API - the order of inputs matters! If you change it, you might lose points.

Use this function to compute the values that you get from taking the mean of the first buff buses' lateness, the first two buff buses' lateness, the first three buff buses' lateness, and so on up to all of the `bus` data points from **Part D**. Store your streaming bus means in a numpy array called `buffbus_bad_means`.  Report all 12 estimates in `buffbus_bad_means`.

In [11]:
# Given API:
def update_mean(prev_mean, xn, n):
    new_mean = prev_mean + (xn - prev_mean)/n
    return(new_mean)

buffbus_bad_means = []

prior_mean = bus[0]

length_of_loop = np.size(bus)
for i in range(length_of_loop):
    prior_mean = update_mean(prior_mean,bus[i],i+1)
    buffbus_bad_means.append(prior_mean)
buffbus_bad_means = np.array(buffbus_bad_means)

for idx,times in enumerate(buffbus_bad_means):
    print("The first {:.0f} busses have a mean lateness of {:.3f} minutes".format(idx+1,times))

The first 1 busses have a mean lateness of 312.000 minutes
The first 2 busses have a mean lateness of 158.000 minutes
The first 3 busses have a mean lateness of 108.667 minutes
The first 4 busses have a mean lateness of 81.500 minutes
The first 5 busses have a mean lateness of 69.600 minutes
The first 6 busses have a mean lateness of 64.500 minutes
The first 7 busses have a mean lateness of 66.857 minutes
The first 8 busses have a mean lateness of 60.875 minutes
The first 9 busses have a mean lateness of 55.000 minutes
The first 10 busses have a mean lateness of 55.500 minutes
The first 11 busses have a mean lateness of 57.727 minutes
The first 12 busses have a mean lateness of 56.417 minutes


To ensure your function complies with the given API, run this small test, where we suppose we have a mean of $\bar{x}_n = 1$ with the first $2$ data points (`prev_mean`), and we update this with the 3rd ($n=3$) data point which is $x_3=2$:

In [13]:
assert update_mean(1,2,3)==4/3, "Warning: function seems broken."

**Part F**:

How many floating point operations were spent computing the final result in your code in part E?  Is this truly better than the uninformed approach from part B?

**Solution:**

The update_mean function took only 3 operations, and we perform it for all data points after the first ($n-1$ times total).  This means that for a data set that streams in $n$ data point, we preform $3(n-1)$ operations compared to the naive approach taking $\frac{n(n+1)}{2}$ operations.  This is an *order* $n$ operation instead of $n^2$, so the larger the data gets, the more time we save!

On our example with only 12 data points, the naive approach would perform 78 operations and the update_mean function just 33, so we're already saving plenty of time.


<a id='Appendix'></a>

## Appendix 

*Goal*: Prove that 
$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$

Note that you can get an expression for $\bar{x}_{n-1}$ by simply replacing $n$ in Equation 1 above with $n-1$.

We'll start with $\bar{x}_n$ and massage it until we get the righthand side of the formula

\begin{eqnarray}
\nonumber \bar{x}_n &=& \frac{1}{n} \sum_{k=1}^n x_k \\
&=& \frac{1}{n} \sum_{k=1}^{n-1} x_k + \frac{1}{n}x_n \\
&=& \frac{n-1}{n-1}\frac{1}{n} \sum_{k=1}^{n-1} x_k + \frac{1}{n}x_n \\
&=& \frac{n-1}{n} \left(\frac{1}{n-1} \sum_{k=1}^{n-1} x_k\right) + \frac{1}{n}x_n \\
&=& \frac{n-1}{n} \bar{x}_{n-1} + \frac{1}{n}x_n \\
&=& \frac{n}{n}\bar{x}_{n-1} - \frac{1}{n}\bar{x}_{n-1} + \frac{1}{n}x_n \\
&=&  \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n} \quad \checkmark
\end{eqnarray}

